### PR TITLE
LexRequirement as a struct, instead of a type

### LINKED ISSUE #12255: making `LexRequirement` an actual struct
```
I have always thought that making `LexRequirement` an actual struct with methods and documentation would be even better

For example,  something like

/// Describes a lexical (multi-column) sort requirement
struct LexRequirement {
  inner: Vec<PhysicalSortRequirement>>
}

impl LexRequirement { 
// documented methods here
...
}

Otherwise the rust docs just show as `Vec<PhysicalSortRequirement>` and it is unclear what operations are used
```

_Originally posted by @alamb in https://github.com/apache/datafusion/pull/12130#pullrequestreview-2257616876_

### PR DESCRIPTION
**Pull request by ngli-me**
## Which issue does this PR close?

Closes #12255.

## What changes are included in this PR?

Converted LexRequirement type into a struct, with the `Vec<PhysicalSortRequirement>` as the field labeled `inner`.
Added two functions for convenience, an `iter` and `push` function relating to the struct field.
Added implementations for `Deref`, `FromIterator` and `IntoIterator`.

## Are these changes tested?

Adjusted the existing tests

## Are there any user-facing changes?

The `Vec<PhysicalSortRequirement>` needs to be wrapped to initialize the struct insance, and some of the usage needs to be adjusted to get the field from the struct.

### CODE REVIEW

**Review by berkaysynnada — APPROVED** [review]
LGTM, thank you @ngli-me

File: datafusion/physical-expr-common/src/sort_expr.rs [hunk 1]
```diff
@@ -264,9 +267,55 @@ pub type LexOrdering = Vec<PhysicalSortExpr>;
 /// a reference to a lexicographical ordering.
 pub type LexOrderingRef<'a> = &'a [PhysicalSortExpr];
```
Comment (berkaysynnada) [inline datafusion/physical-expr-common/src/sort_expr.rs:268]: Perhaps we could follow a similar pattern for LexOrderingRef `too`

**Comment by ngli-me** [comment]
Made some adjustments, since there was a merge conflict :)

File: datafusion/physical-expr-common/src/sort_expr.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr-common/src/sort_expr.rs:268]: BTW having a real struct will make printing these structures much easier -- e.g. I had to do some akward workarounds to implement `Display` or `LexOrdering` in https://github.com/apache/datafusion/pull/12590

**Review by alamb — APPROVED** [review]
Love it -- thank you @ngli-me  and @berkaysynnada

**Pull request merged**